include/deal.II/dofs/dof_accessor.templates.h

```diff
@@ -1588,6 +1588,223 @@ namespace internal
+          dofs_per_quad   = accessor.get_fe(fe_index).dofs_per_quad,   //
+          dofs_per_hex    = accessor.get_fe(fe_index).dofs_per_hex;    //
+
+        const unsigned int inner_dofs =
+          structdim == 1 ? dofs_per_line :
+                           (structdim == 2 ? dofs_per_quad : dofs_per_hex);
+
+        unsigned int index = 0;
+
+        // 1) VERTEX dofs
+        for (const unsigned int vertex :
+             GeometryInfo<structdim>::vertex_indices())
+          for (unsigned int d = 0; d < dofs_per_vertex; ++d, ++index)
+            Operation::process_vertex_dof(
+              accessor, vertex, d, dof_indices[index], fe_index);
+
+        // 2) copy dof numbers from the LINE. for lines with the wrong
+        // orientation (which might occur in 3d), we have already made sure that
+        // we're ok by picking the correct vertices (this happens automatically
+        // in the vertex() function). however, if the line is in wrong
+        // orientation, we look at it in flipped orientation and we will have to
+        // adjust the shape function indices that we see to correspond to the
+        // correct (face/cell-local) ordering.
+        if (structdim == 2 || structdim == 3)
+          for (unsigned int line = 0;
+               line < GeometryInfo<structdim>::lines_per_cell;
+               ++line)
+            for (unsigned int d = 0; d < dofs_per_line; ++d, ++index)
+              Operation::process_dof(
+                *accessor.line(line),
+                accessor.get_fe(fe_index)
+                  .adjust_line_dof_index_for_line_orientation(
+                    d, accessor.line_orientation(line)),
+                dof_indices[index],
+                fe_index);
+
+        // 3) copy dof numbers from the FACE. for faces with the wrong
+        // orientation, we have already made sure that we're ok by picking the
+        // correct lines and vertices (this happens automatically in the line()
+        // and vertex() functions). however, if the face is in wrong
+        // orientation, we look at it in flipped orientation and we will have to
+        // adjust the shape function indices that we see to correspond to the
+        // correct (cell-local) ordering. The same applies, if the face_rotation
+        // or face_orientation is non-standard
+        if (structdim == 3)
+          for (unsigned int quad = 0;
+               quad < GeometryInfo<structdim>::quads_per_cell;
+               ++quad)
+            for (unsigned int d = 0; d < dofs_per_quad; ++d, ++index)
+              Operation::process_dof(
+                *accessor.quad(quad),
+                accessor.get_fe(fe_index)
+                  .adjust_quad_dof_index_for_face_orientation(
+                    d,
+                    accessor.face_orientation(quad),
+                    accessor.face_flip(quad),
+                    accessor.face_rotation(quad)),
+                dof_indices[index],
+                fe_index);
```

Comment:
A side comments regarding hp (unrelated the changes in this PR):  here and above the global `fe_index` is passed, which might lead to the fact that - during the call of this function (inside of the loop) - the local line/quad `fe_index`/offset has to be recompute. I have my doubt that the compiler can optimize it away...

I think it might be an option to precompute the local `fe_index`/offset before the loop.

Reply:
How heavy are the `quad(quad)` or `get_fe(fe_index)` functions that could be moved out. I agree that the compiler won't easily optimize it because we're potentially modifying the object it operates on. In any case, I would not bother too much unless it shows up in profilers.

Reply:
So maybe it is best to simply leave a comment in case someone comes back to this function?

Reply:
Let's rather open an issue. I see `quad()`/`line()` as a potential problem since these have to go through multiple indirections due to re-orientations. I bigger problem is the `fe_index` because it might to have to perform a search at a lot of points. But let's investigate once the merged `DoFHandler` is merged!